File: src/razie/Log.scala
/**  ____    __    ____  ____  ____,,___     ____  __  __  ____
 *  (  _ \  /__\  (_   )(_  _)( ___)/ __)   (  _ \(  )(  )(  _ \           Read
 *   )   / /(__)\  / /_  _)(_  )__) \__ \    )___/ )(__)(  ) _ <     README.txt
 *  (_)\_)(__)(__)(____)(____)(____)(___/   (__)  (______)(____/    LICENSE.txt
 */
package razie
 import com.razie.pub.base.{log => pblog}

/** 
 * some logging basics 
 * 
 * You can customize the entire logging system by providing an implementation for 
 * com.razie.pub.base.log.Factory and set that: com.razie.pub.base.log.Log.factory = new MyFactory().
 * 
 * See example in razpub:com.razie.pub.base.log.Log4j
 * 
 * @author razvanc
 */
trait Log {
   def trace (f : => Any) 
   def log   (msg:String, e:Throwable=null)
   def alarm (msg:String, e:Throwable=null) // implicit audit
   def audit (msg:String, t:Throwable=null)
   def error (msg:String, t:Throwable=null) // alarm and throw
   
   def apply (msg:String, e:Throwable=null) = log (msg, e)
} 

class RaziepbLog extends Log {
   override def trace (f : => Any) = pblog.Log.traceThis ({val x = f; x.toString}) 
   override def log   (msg:String, t:Throwable=null) = pblog.Log.logThis (msg, t)
   override def alarm   (msg:String, t:Throwable=null) = pblog.Log.alarmThis (msg, t)
   override def audit   (msg:String, t:Throwable=null) = pblog.Log.logThis ("AUDIT:"+msg, t)
   override def error   (msg:String, t:Throwable=null) = pblog.Log.alarmThisAndThrow(msg, t)
}

class StupidLog extends Log {
   override def trace (f : => Any) = println ("DEBUG: " + f) 
   override def log   (msg:String, t:Throwable=null) = println ("LOG: " + msg + t)
   override def alarm   (msg:String, t:Throwable=null) = println ("ALARM: " + msg, t)
   override def audit   (msg:String, t:Throwable=null) = println ("AUDIT: " + msg, t)
   override def error   (msg:String, t:Throwable=null) = { println(msg, t); throw t }
}

/** some logging basics 
 * 
 * @author razvanc
 */
object Log extends Log {
   // overwrite/change this to use different logging mechanism
   var impl = new RaziepbLog()
   
   override def trace (f : => Any) = impl.trace(f)
   override def log   (msg:String, t:Throwable=null) = impl.log(msg, t)
   override def alarm   (msg:String, t:Throwable=null) = impl.alarm(msg, t)
   override def audit   (msg:String, t:Throwable=null) = impl.audit(msg, t)
   override def error   (msg:String, t:Throwable=null) = impl.error(msg, t)

   def auditThis   (msg:String, t:Throwable=null) =
      if (t == null)
         pblog.Log.audit (msg)
      else
         pblog.Log.audit (msg, t)

   def logThis (msg:String) = log (msg)
   def apply   (msg:String) = log (msg)
   def logThis (msg:String, e:Throwable) = log(msg, e)
   
   /** @return the same message, so you can return it */
   def alarmThis (msg:String) = {
      impl.alarm(msg) 
      msg
	   }
	   
   /** @return the same message, so you can return it */
   def alarmThis (msg:String, e:Throwable) = {
      impl.alarm (msg, e) 
      msg
	   }

   /** optimized so the code is not even invoked if tracing is off ... don't suppose this will cause side-effects? */
   def traceThis (f : => Any) = {
      if (pblog.Log.isTraceOn()) {
         val p = f
         p match {
            case s:String => pblog.Log.traceThis (s) 
            case (s:String,e:Throwable) => pblog.Log.traceThis (s,e) 
            case _ => impl.trace(p.toString)
         }
      }
   }
}

object Debug {
   def apply (f : => Any) = Log.traceThis (f)
}

object Audit {
   def apply (f : => Any) = Log.audit (f.toString)
}

object Alarm {
   def apply (f : => Any) = Log.alarm (f.toString)
}

object Error {
   def apply (f : => Any) = Log.error (f.toString)
}

File: src/com/razie/pub/base/log/Log.java
/**
 * Razvan's public code. Copyright 2008 based on Apache license (share alike) see LICENSE.txt for
 * details. No warranty implied nor any liability assumed for this code.
 */
package com.razie.pub.base.log;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * simple log proxy - log4j is dominant but then there's the JDK's log... this gives you the freedom
 * to use one or the other...or simply recode to use your own - if you hapen to use another
 * one...doesn't it suck when you use a library which writes to stdout?
 * 
 * create logs per class with the Factory. then invoke log() trace() or alarm().
 * 
 * if you're lazy, use the static Log.logThis()
 * 
 * @author razvanc99
 */
public class Log {

   private String         category;
   private String         component;
   public static String   program    = "dflt";
//   public static int      MAXLOGS    = 1000;
//   public static String[] lastLogs   = new String[MAXLOGS];
//   public static int      curLogLine = 0;
   public static boolean  DEBUGGING  = true;

   public Log(String componentNm, String categoryNm) {
      this.category = categoryNm;
      this.component = componentNm;
   }

   /** from http://www.javapractices.com/topic/TopicAction.do?Id=78 */
   public static String getStackTraceAsString(Throwable aThrowable) {
       final Writer result = new StringWriter();
       final PrintWriter printWriter = new PrintWriter(result);
       aThrowable.printStackTrace(printWriter);
       return result.toString();
     }

   public void log(String m, Throwable t) {
      log(m + " Exception: " + getStackTraceAsString(t));
   }

   public void log(Object... o) {
      String m = "";
      for (int i = 0; i < o.length; i++) {
         m += o[i].toString();
      }

      String msg = "LOG-" + program + "-" + component + "-" + category + ": " + m;
      System.out.println(msg);
   }

   public void alarm(String m, Throwable... e) {
      log(m + (e.length <= 0 ? "" : getStackTraceAsString(e[0])));
   }

   /**
    * trace by concatenating the sequence of objects to String - this is the most efficient trace
    * since the strings will only be evaluated and concatenated if the trace is actually turned on
    */
   public void trace(int l, Object... o) {
      if (isTraceLevel(l)) {
         String m = "";
         for (int i = 0; i < o.length; i++) {
            m += (o[i] == null ? "null" : o[i].toString());
         }
         log(m);
      }
   }

   public boolean isTraceLevel(int l) {
      return DEBUGGING;
   }

   public static boolean isTraceOn() {
      return DEBUGGING;
   }

   public static void logThis(String m) {
      logger.log(m);
   }

   // TODO 2-1 implement the separate audit facility
   public static void audit(String m) { logger.log("AUDIT: "+m); }
   public static void audit(String m, Throwable t) { logger.log("AUDIT: "+m, t); }

   public static void traceThis(String m) { logger.trace(1, m); }
   public static void traceThis(String m, Throwable t) { logger.trace(1, m, t); }

   public static void logThis(String m, Throwable t) {
//      Factory.logger.log(m + " Exception: " + Exceptions.getStackTraceAsString(t));
      logger.log(m + " Exception: " + t);
   }

//   /** alarm this only once in this run...*/
//   public static void alarmOnce(String errorcode, String m, Throwable... e) {
//      if (! alarmedOnce.containsKey(errorcode)) {
//         Factory.logger.alarm(m, e);
//         alarmedOnce.put(errorcode, errorcode);
//      }
//   }

   /** alarm this */
   public static void alarmThis(String m, Throwable... e) {
      logger.alarm(m, e);
   }

   /** alarm this and throw a new RT exception with the message and the cause */
   public static void alarmThisAndThrow(String m, Throwable... e) {
      // TODO i don't think this should log again...since it throws it, eh?
      logger.alarm(m, e);
      if (e.length > 0 && e[0] != null)
         throw new RuntimeException(m, e[0]);
      else
         throw new RuntimeException(m);
   }

//   /**
//    * helper to turn lists/arrays/maps into strings for nice logging
//    * 
//    * @param ret object to toString
//    * @return either the new String or the original object if not recognized
//    */
//   @SuppressWarnings("unchecked")
//   public static Object tryToString(String indent, Object ret) {
//      return Log4j.tryToString(indent, ret);
//   }

   public static Log logger = Factory.create("?", "DFLTLOG");
   
   public static Log create(String component, String categoryName) {
      return factory.create(component, categoryName);
   }

   public static Log create(String categoryName) {
      return factory.create( categoryName);
   }
   
   public static Factory factory = new Factory();
}
